From 0ced2f84ec8fcf961a1d89df1194634d5f83d4fd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Diego=20Elio=20'Flameeyes'=20Petten=C3=B2?= <flameeyes@gmail.com>
Date: Mon, 15 Mar 2010 18:51:50 +0100
Subject: [PATCH] Correct the check for expected values returned by the acct query.

The current code checked for at least two rows and at most three columns,
while the correct check was to be one row, and between two and three
columns.

This was reported by Gianpaolo Tomassoni in Gentoo bug #309311
http://bugs.gentoo.org/show_bug.cgi?id=309311
---
 src/pam_pgsql.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/pam_pgsql.c b/src/pam_pgsql.c
index 5a31ac1..8126b79 100644
--- a/src/pam_pgsql.c
+++ b/src/pam_pgsql.c
@@ -129,7 +129,8 @@ pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
 					DBGLOG("query: %s", options->query_acct);
 					rc = PAM_AUTH_ERR;
 					if(pg_execParam(conn, &res, options->query_acct, pam_get_service(pamh), user, NULL, rhost) == PAM_SUCCESS) {
-						if (PQntuples(res) >= 2 && PQnfields(res) <= 3) {
+						if (PQntuples(res) == 1 &&
+						    PQnfields(res) >= 2 && PQnfields(res) <= 3) {
 							char *expired_db = PQgetvalue(res, 0, 0);
 							char *newtok_db = PQgetvalue(res, 0, 1);
 							rc = PAM_SUCCESS;
@@ -143,7 +144,7 @@ pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
 							if (!strcmp(expired_db, "t"))
 								rc = PAM_ACCT_EXPIRED;
 						} else {
-							DBGLOG("query_acct should return two or three columns");
+							DBGLOG("query_acct should return one row and two or three columns");
 							rc = PAM_PERM_DENIED;
 						}
 						PQclear(res);
-- 
1.7.0.2

